### DISCUSSION TITLE
Saving human labels from .pkg.slp or .slp file

### DISCUSSION THREAD

**Discussion by catubc** [talmolab/sleap#1405] [Help!]
Hello 

We are looking to release a large video dataset for public use in machine vision tasks (likely as an eLife Resource paper).  As part of the data release, we'd like to provide our human annotated frames from existing .pkg.slp or .slp file so that other labs/users can use the datasets in their machine vision pipelines (that may not be based on sleap).  That is, we want to make the ground truth labels available in a simple format that anyone can use if they seek to develop methods that work on the larger video dataset.

Is there a simple way to extract the frames + human annotations from the .pkg.slp or .slp files to a simple format for general use?  

I would assume that an end user would want some sort of tuple dictionary containing {the labels, the individual video frame}.  I'm not sure if sleap already provides for that - otherwise we can try our own code from scratch.

Any advice?

Thanks so much,
catubc

**Comment by talmo** [comment]
Hey @catubc,

Yep, full SLEAP is an annoying dependency for just being able to read `.slp` files since installing TensorFlow and everything else is a pain. For this reason, we're working on [sleap-io](https://github.com/talmolab/sleap-io) which is a standalone library for just that.

**sleap-io** is pip installable and has minimal dependencies, making it pretty compatible with most settings. That said, you might want to archive data in a format that doesn't require _any_ deps at all. It's hard to do this without having a super standard data format, especially given that this field is rapidly advancing, resulting in changing needs and capabilities for such a data format.

The answer is flexibility, which sleap-io aims to enable. While we work to add support for different formats, [here's a notebook example of how you could use it](https://colab.research.google.com/drive/1WVCuwKfc2-icqYL0VJFj9dvCGCB5z9Fn). In that notebook, we go from `.pkg.slp` files to the [Ultralytics pose dataset format](https://docs.ultralytics.com/datasets/pose/) which is basically folders of images with pose annotations in text files.

There will also soon be an [update to the NWB pose format](https://github.com/rly/ndx-pose/pull/9#issuecomment-1645162259) which will better support storing data for training (i.e., random images rather than timeseries) which would be a good fit for this.

Let me know if you have any questions or if there's anything we can help with!

Talmo

**Comment by catubc** [comment]
Hi @talmo 

So we've run the code, thanks so much for that. The only changes we made were in the filenames/structures:

```
dataset_path = "/mnt/b3a68699-495d-4ebb-9ab1-ac74f11c68c5/sleap/io_tests/"
dataset_name = "gerbils6"

labels_paths = {
    "train": "/mnt/b3a68699-495d-4ebb-9ab1-ac74f11c68c5/sleap/io_tests/cohort1_day_subsample_100.pkg.slp",
    #"val": "D:\sleap-data\datasets\wt_gold.13pt\tracking_split2\val.pkg.slp",
    #"test": "D:\sleap-data\datasets\wt_gold.13pt\tracking_split2\test.pkg.slp",
}

symmetries = []

class_index = 0
quality = 90
overwrite = True
```

And it started generating the data which looks like this (this is for 6 animal videos with 6 features each):
```

0 0.530802 0.504494 0.152965 0.027408 0.454319 0.490790 1 0.479552 0.510911 1 0.510101 0.517144 1 0.544250 0.517953 1 0.577761 0.516808 1 0.607284 0.518198 1
0 0.836902 0.754532 0.111093 0.049961 0.781356 0.730419 1 0.821764 0.755231 1 0.851438 0.774847 1 0.877323 0.779513 1 0.892449 0.759329 1 0.891201 0.729551 1
0 0.751431 0.418659 0.096381 0.071140 0.703241 0.454229 1 0.728593 0.423091 1 0.744080 0.402658 1 0.763274 0.386166 1 0.783232 0.383089 1 0.799621 0.397879 1
0 0.793715 0.309928 0.055153 0.129770 0.821292 0.374814 1 0.814905 0.324748 1 0.808423 0.302187 1 0.797424 0.278746 1 0.784163 0.258501 1 0.766139 0.245043 1
0 0.919993 0.446333 0.106648 0.088216 0.973318 0.402225 1 0.962503 0.432917 1 0.950282 0.463491 1 0.925342 0.485173 1 0.892401 0.490441 1 0.866669 0.459482 1
0 0.842690 0.277361 0.058915 0.123932 0.813232 0.215395 1 0.850000 0.242823 1 0.859609 0.256716 1 0.866061 0.276926 1 0.872147 0.303133 1 0.871287 0.339327 1
``` 
Does this look about right?  It seems there are 4 starting float points and then triplets containing x,y,visibility?

But we also got a crash before the .yaml file could be created (this occurred after all 1000 labeled frames were generated):

```
/home/cat/miniconda3/envs/sleap_new/lib/python3.7/site-packages/rich/live.py:221: 
UserWarning: install "ipywidgets" for Jupyter support
  warnings.warn('install "ipywidgets" for Jupyter support')

/home/cat/miniconda3/envs/sleap_new/lib/python3.7/site-packages/ipykernel_launcher.py:24: 
RuntimeWarning: All-NaN slice encountered

/home/cat/miniconda3/envs/sleap_new/lib/python3.7/site-packages/ipykernel_launcher.py:25: 
RuntimeWarning: All-NaN slice encountered

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_16336/122217950.py in <module>
     10     img_folder = dataset_path / dataset_name / split_name
     11 
---> 12     convert_frames(labels, save_folder=img_folder, class_index=class_index, quality=quality, overwrite=overwrite)
     13 
     14 write_dataset_yaml(dataset_path, dataset_name, n_nodes, flip_idx)

/tmp/ipykernel_16336/2689248743.py in convert_frames(labels, save_folder, class_index, quality, overwrite)
     79             continue
     80 
---> 81         img = lf.image.squeeze()
     82         img_height, img_width = img.shape[:2]
     83         instances = "\n".join([

~/miniconda3/envs/sleap_new/lib/python3.7/site-packages/sleap_io/model/labeled_frame.py in image(self)
     66     def image(self) -> np.ndarray:
     67         """Return the image of the frame as a numpy array."""
---> 68         return self.video[self.frame_idx]
     69 
     70     @property

~/miniconda3/envs/sleap_new/lib/python3.7/site-packages/sleap_io/model/video.py in __getitem__(self, inds)
    131                 "automatically from the filename."
    132             )
--> 133         return self.backend[inds]

~/miniconda3/envs/sleap_new/lib/python3.7/site-packages/sleap_io/io/video.py in __getitem__(self, ind)
    233         """
    234         if np.isscalar(ind):
--> 235             return self.get_frame(ind)
    236         else:
    237             if type(ind) is slice:

~/miniconda3/envs/sleap_new/lib/python3.7/site-packages/sleap_io/io/video.py in get_frame(self, frame_idx)
    177         See also: `get_frames`
    178         """
--> 179         img = self._read_frame(frame_idx)
    180 
    181         if self.grayscale is None:

~/miniconda3/envs/sleap_new/lib/python3.7/site-packages/sleap_io/io/video.py in _read_frame(self, frame_idx)
    542 
    543             if self.frame_map:
--> 544                 frame_idx = self.frame_map[frame_idx]
    545 
    546             img = ds[frame_idx]

KeyError: 432

```

Thanks again for the quick responses.

**Comment by catubc** [comment]
I bypassed the error (using try/catch) to the yaml file and it now generated a .yaml file:

```
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: cohort2  # dataset root dir
train: train  # train images (relative to 'path')
val: val  # val images (relative to 'path')
test: test  # test images (optional)

# Keypoints
kpt_shape: [6, 3]  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
flip_idx: [0, 1, 2, 3, 4, 5]

# Classes dictionary
names:
  0: instance
```

Does this seem also about right? I checked the Ultralytics and seems pretty consistent with what your script generated.  From our end missing 1 or 2 frames is fine for downstream users, but I'm not sure if something else would need to be done.

If so, then thanks so much!  Feel free to close the issue.

**Comment by talmo** [comment]
Hi @catubc,

> Does this look about right? It seems there are 4 starting float points and then triplets containing x,y,visibility?

Yep, the [Ultralytics pose dataset format](https://docs.ultralytics.com/datasets/pose/) format saves annotations in:

```
<class-index> <x> <y> <width> <height> <px1> <py1> <p1-visibility> <px2> <py2> <p2-visibility> <pxn> <pyn> <p2-visibility>
```

> But we also got a crash before the .yaml file could be created (this occurred after all 1000 labeled frames were generated):

Very odd! This looks like it's trying to read a frame that isn't in the .pkg.slp file. It's possible that there are additional labeled frames in there that don't have corresponding embedded images. These might be frames with only predictions/suggestions perhaps?

We could track that down, but as long as you catch it in a try-except so that you don't save corresponding annotations with missing images it's probably fine.

I'm more concerned about this message:

```
RuntimeWarning: All-NaN slice encountered
```

This suggests that there are instances with all points not visible, which may be generating annotations that are invalid. I'm not sure why that would be happening in the first place, but I'd add another check for that and skip instances that are all `NaN`. This is more of a data integrity issue that should be double checked, but the quick fix would be to change this:

```py
        instances = "\n".join([
            convert_instance(instance, img_height, img_width, class_index=class_index)
            for instance in lf.user_instances
        ])
```

to this:

```py
    instances = []
    for instance in lf.user_instances:
        if np.isnan(instance.numpy()).all():
            print(f"Skipping empty instance on {lf.video} -> frame {lf.frame_idx}.")
        else:
            instances.append(convert_instance(instance, img_height, img_width, class_index=class_index))
    instances = "\n".join(instances)
```

It'll catch those to prevent writing out bad instances, as well as printing out where there might be issues with the annotations so that you can check them manually.
